test/integration/throttleQueueTest.ts

```diff
@@ -0,0 +1,161 @@
+import { strict as assert } from "assert";
+
+import { UserID } from "matrix-bot-sdk";
+import { ThrottlingQueue } from "../../src/queues/ThrottlingQueue";
+
+describe("Test: ThrottlingQueue", function() {
+    it("Tasks enqueued with `push()` are executed exactly once and in the right order", async function() {
+        this.timeout(20000);
+
+        const queue = new ThrottlingQueue(this.mjolnir, 10);
+        let state = new Map();
+        let promises = [];
+        for (let counter = 0; counter < 10; ++counter) {
+            const i = counter;
+            promises.push(queue.push(async () => {
+                if (state.get(i)) {
+                    throw new Error(`We shouldn't have set state[${i}] yet`);
+                }
+                state.set(i, true);
+                for (let j = 0; j < i; ++j) {
+                    if (!state.get(j)) {
+                        throw new Error(`We should have set state[${j}] already`);
+                    }
+                }
+            }));
+        }
+        await Promise.all(promises);
+        // Give code a little bit more time to trip itself.
```

Comment:
Hmm? But i thought that the promise returned by `push` would be resolved if the task is complete?